File: hex_terrain_editor/editor/eraser_tool.py
"""Eraser tool for removing decorators from hexes."""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from ..utils.hex_math import pixel_to_hex

if TYPE_CHECKING:
    from ..model.hex_grid import HexGrid
    from .tool_manager import ToolManager


class EraserTool:
    """Click or drag to remove all decorators from hexes."""

    def __init__(
        self,
        grid: "HexGrid",
        tool_manager: "ToolManager",
        on_hex_changed: Any = None,
    ) -> None:
        self.grid = grid
        self.tool_manager = tool_manager
        self.on_hex_changed = on_hex_changed
        self._erasing = False
        self._old_states: dict[tuple[int, int], dict] = {}

    def activate(self) -> None:
        pass

    def deactivate(self) -> None:
        self._erasing = False

    def on_press(self, world_x: float, world_y: float) -> None:
        self._erasing = True
        self._old_states = {}
        self._erase_at(world_x, world_y)

    def on_move(self, world_x: float, world_y: float) -> None:
        if self._erasing:
            self._erase_at(world_x, world_y)

    def on_release(self, world_x: float, world_y: float) -> None:
        if self._erasing and self._old_states:
            new_states = {}
            for key in self._old_states:
                cell = self.grid.get(key[0], key[1])
                if cell:
                    new_states[key] = cell.to_dict()

            self.tool_manager.push_undo(
                "Erase decorators",
                {"cells": self._old_states},
                {"cells": new_states},
            )
        self._erasing = False
        self._old_states = {}

    def _erase_at(self, world_x: float, world_y: float) -> None:
        q, r = pixel_to_hex(world_x, world_y, self.grid.hex_size)
        cell = self.grid.get(q, r)
        if cell is None or not cell.decorators:
            return

        key = (q, r)
        if key not in self._old_states:
            self._old_states[key] = cell.to_dict()

        cell.decorators.clear()
        if self.on_hex_changed:
            self.on_hex_changed(q, r)

File: src/editor/eraser_tool.rs
use std::collections::HashMap;
use crate::hex_math;
use crate::model::hex_cell::HexCell;
use crate::model::hex_grid::HexGrid;
use super::tool_manager::{ToolManager, UndoEntry};

pub struct EraserTool {
    pub radius: i32,
    erasing: bool,
    old_states: HashMap<(i32, i32), HexCell>,
}

impl EraserTool {
    pub fn new() -> Self {
        Self {
            radius: 0,
            erasing: false,
            old_states: HashMap::new(),
        }
    }

    pub fn on_press(
        &mut self,
        world_x: f32,
        world_y: f32,
        grid: &mut HexGrid,
        tool_manager: &mut ToolManager,
    ) {
        self.erasing = true;
        self.old_states.clear();
        self.erase_at(world_x, world_y, grid, tool_manager);
    }

    pub fn on_move(
        &mut self,
        world_x: f32,
        world_y: f32,
        grid: &mut HexGrid,
        tool_manager: &mut ToolManager,
    ) {
        if self.erasing {
            self.erase_at(world_x, world_y, grid, tool_manager);
        }
    }

    pub fn on_release(&mut self, tool_manager: &mut ToolManager, grid: &HexGrid) {
        if !self.erasing {
            return;
        }
        self.erasing = false;

        if self.old_states.is_empty() {
            return;
        }

        let mut new_states = HashMap::new();
        for &(q, r) in self.old_states.keys() {
            if let Some(cell) = grid.get(q, r) {
                new_states.insert((q, r), cell.clone());
            }
        }

        tool_manager.push_undo(UndoEntry {
            name: "Erase Decorators".to_string(),
            old_cells: std::mem::take(&mut self.old_states),
            new_cells: new_states,
        });
    }

    fn erase_at(
        &mut self,
        world_x: f32,
        world_y: f32,
        grid: &mut HexGrid,
        tool_manager: &mut ToolManager,
    ) {
        let (cq, cr) = hex_math::pixel_to_hex(world_x, world_y, grid.hex_size);
        let affected = hex_math::hexes_in_radius(cq, cr, self.radius);

        for (q, r) in affected {
            if let Some(cell) = grid.get_mut(q, r) {
                if cell.decorators.is_empty() {
                    continue;
                }
                if !self.old_states.contains_key(&(q, r)) {
                    self.old_states.insert((q, r), cell.clone());
                }
                cell.decorators.clear();
                tool_manager.mark_dirty(q, r);
            }
        }
    }
}

impl Default for EraserTool {
    fn default() -> Self {
        Self::new()
    }
}
